### DISCUSSION TITLE
Custom Id values

### DISCUSSION THREAD

**Discussion by kj4bix** [Azure/azure-mobile-apps#671] [General]
I have an existing database using the old Azure Mobile Client SDK. I'm now moving to the new Azure mobile sdk using Offline tables. In my current database I used a custom Id column that is a primary key to the data in some of my tables. The value is unique and works fine in the older mobile application. However, using the new SDK I'm getting an error that my Id is invalid. Is there someway to tell it that my Id value is not a default value and is ok? I hope it isn't the | causing the issue, this worked in the older SDK. 

This is from the exception message '22-5430|571972766562' is an invalid ID (Parameter 'instance').

This exception is on the client, the server seems to be fine with the value. 

Thanks

**Comment by adrianhall** [comment]
The definition of a valid ID is here: https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client/Utils/Arguments.cs#LL35C9-L35C96

``` csharp
private static readonly Regex validIdRegex = new("^[a-zA-Z0-9][a-zA-Z0-9-_.]{0,126}$");
```

So, yes, the | is not allowed.  Having said that, there is no reason why we can't expand the definition of a valid ID.  Submit an issue, and I'll work on it in the next release.

**Reply by kj4bix** [reply]
Is there anyway I can work around this to get the application running?

**Reply by adrianhall** [reply]
Download the source; add the | symbol to the list of allowed characters in the Arguments.cs, compile and use your version of the library.

**Comment by kj4bix** [comment]
Adrian - is there anyway a path for bypassing the ID validation is possible. This is the only change I have in my version and would really like to use the real package. In Arguments.cs, IsValidId, I still have the null check, but commented the regular expression check. I don't know the best way to "weave" a setting into Arguments. I think most realize they need a unique ID and having the regex meet all the necessary values can be difficult. BTW - my app is now in production in the stores and it working really well. Thanks for your help. 

Alan

**Reply by adrianhall** [reply]
The ID must meet the following criteria:

* A non-null string
* Valid as the segment of a HTTP path without encoding
* Globally Unique

As I said, I'm happy to expand the regular expression, but let's focus it on what is actually required.  That second bullet is probably what is killing things.  I'll expand the regular expression to include the definition here: https://datatracker.ietf.org/doc/html/rfc3986#section-3.3 - except for the %encodings (which aren't allowed with the code we have and would be hell to support right now).

Having said that, it won't be until we switch over to CommunityToolkit (and I'm starting that work tomorrow), so don't expect it quickly.

**Reply by kj4bix** [reply]
One of my id strings while being unique also has spaces (a company name for example). I kept expanding the regex then just finally eliminated it in my code because I knew I had valid ids. I didn't have an encoding issue.

**Reply by adrianhall** [reply]
A space is NOT a valid character for a path segment according to RFC3986.  It needs to be encoded (which is the thing we are trying to avoid here).  I'm going to be VERY resistant to including space in the allowed characters list.  While it works for you (and that's great), it's not something that can be generalized over all the platforms we support.

So this is one of those times where you want to fork the repository and create your own library.

**Reply by kj4bix** [reply]
ok, however it did work in the old version with no issues.
